File: sem2/hw3/task3_1/ru/liamin/vladimir/List.java
package ru.liamin.vladimir;

import java.util.DuplicateFormatFlagsException;

/** Implementation of a unique list */
public class List {
    private int size;
    private ListElement head;

    /**
     * Writing elements to the list
     * @param value value of an element
     * @param currentSize amount of elements
     * @return new amount of elements
     */
    public int push(int value, int currentSize) {

        if (size == 0) {
            head = new ListElement(value, head);
            size++;
            return ++currentSize;
        }

        ListElement temp = head;

        for (int i = 0; i < size - 1; i++) {
            if (temp.value == value){
                try {
                    throw new DuplicateFormatFlagsException("Also");
                }
                finally {
                    System.out.println("Has also");
                    return currentSize;
                }
            }
            temp = temp.next;
        }

        if (temp.value == value)
        {
            try {
                throw new DuplicateFormatFlagsException("Also");
            }
            finally {
                System.out.println("Has also");
                return currentSize;
            }

        }
        size++;
        temp.next = new ListElement(value, temp.next);
        return ++currentSize;
    }

    /**
     * Delete an element in list
     * @param number value of an element
     * @param currentsize amount of elements
     * @return new amount of elements
     */
    public int delete(int number, int currentsize) {
        ListElement temp = head;
        if (size == 0){
            System.out.println("Not found");
            return currentsize;
        }

        if (head.value == number){
            head = head.next;
            size--;
            return currentsize - 1;
        }

        while ((temp.next != null) && (temp.next.value != number))
            temp = temp.next;

        if (temp.next == null) {
            System.out.println("Not found");
            return currentsize;
        }
        temp.next = temp.next.next;
        size--;
        return currentsize - 1;
    }

    /**
     * Check for emptiness
     * @return return true if list is empty
     */
    public boolean isEmpty() {
        return head == null;
    }

    /**
     * Count of elements in list
     * @return number of elements in list
     */
    public int count() {
        return size;
    }

    /** Print list element */
    public void printList() {
        System.out.println("Elements of list: ");
        ListElement current = head;
        while (current != null) {
            System.out.println(current.value);
            current = current.next;
        }
    }

    /**
     * Looks for an item in the list
     * @param number value of an element
     * @return true if an element in list
     */
    public boolean find(int number){
        for (int i = 0; i < size; i++){
            if (head.value == number)
                return true;
        }
        return false;
    }

    /**
     * Returns the value and deletes the element
     * @return number value of an element
     */
    public int pop(){
        if (head != null){
            int value = head.value;
            head = head.next;
            return value;
        }
        return 0;
    }

    /** Clear list */
    public void clear() {
        size = 0;
        head = null;

    }

    private class ListElement {
        private int value;
        private ListElement next;

        public ListElement(int number, ListElement next) {
            value = number;
            this.next = next;
        }
    }
}

File: sem1/test1/task1.1/list.cpp
#include <stdio.h>
#include "list.h"

List *createList() {
    List *list = new List;
    list->first = nullptr;
    list->last = nullptr;
    return list;
}

void push(ListElement *&first, ListElement *&last, int number) {
    while (first != nullptr)
        first = first->next;
    first = new ListElement;
    first->prev = last;
    first->next = nullptr;
    first->value = number;
    if (last != nullptr)
        last->next = first;
    last = first;
    while (first->prev != nullptr)
        first = first->prev;
}

void push(List *&list, int number) {
    push(list->first, list->last, number);
}

bool checkForSymmetry(ListElement *first, ListElement *last) {
    while (first != last) {
        if (first->value != last->value)
            return false;

        last = last->prev;
        first = first->next;
    }
    return true;
}

bool checkForSymmetry(List *list) {
    checkForSymmetry(list->first, list->last);
}

void deleteList(ListElement *first) {
    while (first) {
        ListElement *current = first;
        first = first->next;
        delete current;
    }
}

void deleteList(List *list) {
    deleteList(list->first);
    delete list;
}

void printList(ListElement *first) {
    printf("List: ");
    while (first != nullptr) {
        printf(" %d", first->value);
        first = first->next;
    }
    printf("\n");
}

void printList(List *list) {
    printList(list->first);
}